src/sdl2_compat.c

```diff
@@ -3914,6 +3914,231 @@ DECLSPEC void SDLCALL SDL_SIMDFree(void *ptr)
+    cvt->len_mult = 1;
+    cvt->len_ratio = 1.0;
+    cvt->rate_incr = ((double)dst_rate) / ((double)src_rate);
+
+    /* Use the filters[] to store some data ... */
+    {
+        AudioParam *ptr = (AudioParam *)&cvt->filters[1];
+        cvt->filters[0] = NULL;
+
+        ptr->src_format = src_format;
+        ptr->src_channels = src_channels;
+        ptr->src_rate = src_rate;
+        ptr->dst_format = dst_format;
+        ptr->dst_channels = dst_channels;
+        ptr->dst_rate = dst_rate;
+
+        cvt->needed = 1;
+        if (src_format == dst_rate && src_rate == dst_rate && src_channels == dst_channels) {
+            cvt->needed = 0;
+        }
+ 
+        if (src_rate < dst_rate) {
+            const int mult = (dst_rate / src_rate);
+            cvt->len_mult *= mult;
+            cvt->len_ratio *= mult;
+        } else {
+            const int div = (src_rate / dst_rate);
+            cvt->len_ratio /= div;
+        }
+       
+        if (src_channels < dst_channels) {
+            cvt->len_mult = ((cvt->len_mult * dst_channels) + (src_channels - 1)) / src_channels;
+        }
+
+        /* !!! FIXME: remove this if we can get the resampler to work in-place again. */
+        /* the buffer is big enough to hold the destination now, but
+           we need it large enough to hold a separate scratch buffer. */
+        cvt->len_mult *= 2;
+        
+        cvt->filters[0] = NULL;
+
+    }
+    SDL3_DestroyAudioStream(stream);
+    return cvt->needed;
+}
+
+
+DECLSPEC int SDLCALL SDL_ConvertAudio(SDL_AudioCVT *cvt)
+{
+
+    SDL_AudioStream *stream;
+    Uint8 *dst_buf = NULL;
+    SDL_AudioFormat src_format, dst_format;
+    int src_channels, src_rate;
+    int dst_channels, dst_rate;
+
+    int src_len, dst_len, real_dst_len;
+    int src_samplesize, dst_samplesize;
+ 
+    /* Sanity check target pointer */
+    if (cvt == NULL) {
+        return SDL_InvalidParamError("cvt");
+    }
+   
+    { 
+        AudioParam *ptr;
+        ptr = (AudioParam *)&cvt->filters[1];
+        src_format = ptr->dst_format;
+        src_channels = ptr->src_channels;
+        src_rate = ptr->src_rate;
+        dst_format = ptr->dst_format;
+        dst_channels = ptr->dst_channels;
+        dst_rate = ptr->dst_rate;
+    }
+
+    stream = SDL3_CreateAudioStream(src_format, src_channels, src_rate,
+                                    dst_format, dst_channels, dst_rate);
+    if (stream == NULL) {
+        goto failure;
+    }
+
+    src_samplesize = (SDL_AUDIO_BITSIZE(src_format) / 8) * src_channels;
+    dst_samplesize = (SDL_AUDIO_BITSIZE(dst_format) / 8) * dst_channels;
+
+    src_len = cvt->len & ~(src_samplesize - 1);
+    dst_len = dst_samplesize * (src_len / src_samplesize);
+    if (src_rate < dst_rate) {
+        const double mult = ((double)dst_rate) / ((double)src_rate);
+        dst_len *= (int) SDL_ceil(mult);
+    }
+
+    dst_len = dst_len & ~(dst_samplesize - 1);
+    dst_buf = (Uint8 *)SDL_malloc(dst_len);
```

Comment:
So the AudioCVT struct tells the app how big its buffer needs to be for not just the result, but also for any work the converter needs to do. We shouldn't need to malloc an output buffer here just to memcpy to the actual output buffer, as the app will provide us a buffer as large as we request and we can write directly into it.

Reply:
here, I have some trouble :/
are you sure I can put 
`SDL3_PutAudioStreamData(stream, cvt->buf, src_len)`
then read on the same buffer ?
`real_dst_len = SDL3_GetAudioStreamData(stream, cvt->buf, dst_len) `

because when I tried with `./testresample`, if make a different output wav with has some crackling noise

Reply:
ok. done there was a mistake in my previous code. memcpy was a wrong size.